### DISCUSSION TITLE
Derivative gain in PID not having impact

### DISCUSSION THREAD

**Discussion by ThomasKennings** [JSBSim-Team/jsbsim#673] [Q&A]
EDIT: Solution found! I feel silly now. The issue was that the input to the PID was deliberately being clipped. Once the input value reached the clipping point, its derivative was 0 since it stayed at the clipping point all the time. Removing the clipping shows derivative gain to work. The extreme fluctuations at the beginning weren't an issue after removing the clipping as well.
I'll put the clipping back. It's good to know derivative gain is working.

Hello,

I'm writing a simplified simulation of pilot behavior using the scripting available in the aircraft .xml files. My intention is to use it to hold a given altitude, as well as to adjust to a new altitude when the target altitude variable (new) is changed in a script file. My basic control layout goes something like this:

altitude error calculation -> PID to get vertical speed ref -> vertical speed error calculation -> PID to set elevator-cmd-norm

My issue is that derivative gain in the last PID is having no impact on elevator-cmd-norm. I checked that P and I are working by setting the other two gains to 0.0, and P and I both work. However, setting P=0, I=0, D!=0 causes elevator-cmd-norm to flip between -1 and 1 at 60 Hz for the first ~3.7 seconds, then stay at 0 for the remainder of the simulation.

My derivative gain is 1.0 and my vertical speed error is near 0, so I don't understand the insane oscillation at the start. Then I don't understand why the output is 0.0 while my vertical speed is not zero (plane starts climbing). Does anybody have recommendations or things I should check? 

```xml
    	<pid name="fcs/elevator-cmd-norm">
		<!--make pitch inputs to maintain target vertical speed-->
		<input> fcs/target-vert-speed-diff </input> <!-- in feet per second -->
		<kp> 0.0 </kp>
		<ki> 0.0 </ki>
		<kd> 1.0 </kd>	<!-- derivative gain not working? -->
		<clipto>
	  		<min> -1.0 </min>
	  		<max> 1.0 </max>
		</clipto>
    	</pid>
```

**Comment by seanmcleod** [comment]
@ThomasKennings make use of the triple backtick ` option followed by xml and close it with triple back ticks to format the xml better in future. I've edited your comment.

> My derivative gain is 1.0 and my vertical **speed error is near 0**, so I don't understand the insane oscillation at the start.

Even if the error term is near 0 what counts is what is the derivative, how noisy is the error term? Plot it for the first 3s or so and inspect it and compute the derivative for each time step to see how large the derivative is.

For example, assuming you're using the default 120Hz rate, then let's say 2 samples a single time step apart are say 0.1 and 0.2, then the derivative `(0.2 - 0.1)/0.0083 = 12`, which means given your `kd = 1` the PID output is going to be clipped to 1.

**Reply by ThomasKennings** [reply]
I will take another look to see the things you mentioned. Do you have any idea what may cause the flatline at 0? Would the pilot be simulated as having passed out due to overuse of the control stick? ;)

**Comment by seanmcleod** [comment]
> Do you have any idea what may cause the flatline at 0?

Look at your input data closely, and how you calculate your input data, a PID controller is really very simple.

https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/src/models/flight_control/FGPID.h#L97-L101

https://github.com/JSBSim-Team/jsbsim/blob/8852bf55a3928a940ff1305f95426695095efacc/src/models/flight_control/FGPID.cpp#L164

https://github.com/JSBSim-Team/jsbsim/blob/8852bf55a3928a940ff1305f95426695095efacc/src/models/flight_control/FGPID.cpp#L203
